FEAT: Add login/logout function

B2SIC opened the pull request:
## 관련 Issues
- #6 
## 개요
### 회원 로그인 추가
* `/login` `POST`
  * `LoginForm` -> `userId`, `password` 를 Request Body로 받아서 처리
  * 로그인 성공 -> Session 생성
    * 세션 유지 시간: 30분
    * 마지막 활동 시점으로 세션 시간이 갱신되는 방식
  * 로그인 실패
    * 존재하지 않는 아이디 or 비밀번호 Mismatch -> BusinessException Throw
*  `/logout` `GET`
    * 세션 연결 해제
* 관련 테스트 케이스 추가

B2SIC committed:
FEAT: Add login/logout function
Add login controller
Add login form
Add login service method
Add configuration for session expiration time
Add login test method

namsic commented on src/main/java/harmony/dev/harmonyserver/form/LoginForm.java:
```diff
@@ -0,0 +1,20 @@
+
+@Getter
+@RequiredArgsConstructor
+public class LoginForm {
+    @NotNull
+    private String userId;
+    @NotNull
+    private String password;
```
Member의 DTO inner class로 둡시다.

B2SIC replied:
Member 도메인 안에 inner class로 두라는 뜻인가요? 그렇게 했을 때 얻는 이점이 있나요?

namsic replied:
> Member 도메인 안에 inner class로 두라는 뜻인가요? 그렇게 했을 때 얻는 이점이 있나요?

LoginForm은 userId와 password라는, Member의 필드만으로 구성된
일종의 Request DTO로 볼 수 있기 때문입니다.

https://bestinu.tistory.com/56
https://jaehoney.tistory.com/157?category=983083


추가로, 현재의 ResponseDTO도 목적에 맞게 이름을 변경할 필요가 있어 보입니다.
이후 Member 관련 응답이 다양해지면 ResponseDTO라는 이름은 어느 요청에 대한 응답인지, 어떤 필드로 구성되어 있는지 유추하기 어렵기 때문입니다.
이는 추후 별도 이슈로 생성해 처리하는 것이 좋겠습니다.

B2SIC replied:
로그인 관련 필드로 이루어진 DTO가 존재한다면 해당 DTO 클래스 안에 이너 클래스로 넣어도 좋을 것 같습니다.

그런데 Member 클래스는 DTO가 아닌 도메인 패키지 안에 존재하는 엔티티 클래스인데 이 안에 보관하는 것이 맞을지 잘 모르겠습니다.
특히 내부에서는 데이터베이스 관련 설계 및 처리를 하고 있기 때문에 분리되는 것이 맞다고 생각했습니다.

관련해서 현재 Member 클래스 내에 존재하는 ResponseDto 클래스 역시 나중에는 리팩토링을 통해 다른 클래스로 이전 되거나 분리되어야 한다고 생각합니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/service/MemberService.java:
```diff
@@ -63,4 +61,29 @@ private void validateBeforeSignupMember(Member member) {
+        BusinessException e = new BusinessException();
+
+        Member findUser = null;
+        if (userList.isEmpty()) {
+            e.add(ExceptionSummary.builder()
+                            .field("userId")
+                            .value(userId)
+                            .code("Not Exist")
+                            .build());
+            e.peekaboo();
+        } else {
+            findUser = userList.get(0);
+            if (!findUser.getPassword().equals(password)) {
+                e.add(ExceptionSummary.builder()
+                                .field("password")
+                                .code("Mismatch")
+                                .build());
+                e.peekaboo();
+            }
+        }
```
- 해당 id가 존재하지 않는 경우
- id와 password가 맞지 않는 경우

두 경우의 응답을 구분할 수 없도록 하면 좋겠습니다.
공격자에게 정보가 될 수 있기 때문입니다.

B2SIC replied:
변경 사항에 반영하겠습니다.

namsic reviewed (commented):
이슈에서 논의된대로 URI를 변경하는 것에 대해서도 검토 부탁합니다.

B2SIC committed:
REFACTOR: Modify Member domain and add MemberDTO
Add MemberDTO class
Change validator location (Member -> MemberDTO)
Remove LoginController
Remove LoginForm
Add spring security dependencies and jsonwebtoken
Change lombok version

B2SIC committed:
FEAT: JWT Login function
Add Spring Security Configuration and JWT classes
Add Role Enum
Add auth mapping in MemberController
Add Exception for Not Unique

B2SIC committed:
TEST: Change login test code and test code dependencies

B2SIC commented:
## 리뷰 내용 반영
### `LoginForm` 관련
  * 리뷰에서 제시했던 Member 클래스의 활용하는 방안을 토대로 MemberDTO를 별도로 생성하고 기존 LoginForm은 제거하였습니다.
    * MemberDTO를 별도로 생성한 이유
      * 기존 구조는 Member Entity 클래스를 직접 주고 받는 형태였습니다. 그런데 이러한 구조는 Entity 변경 사항에 유연하게 대응하기 어렵고 역할은 명확하게 구분하는 것이 더 좋은 형태이기 때문에 별도의 객체인 MemberDTO를 사용해서 데이터를 주고 받을 수 있도록 하였습니다.
    * 변경 사항
      * MemberDTO를 별도로 생성함에 따라 기존에 Member 클래스에 있었던 `ResponseDto Class` 를 MemberDTO 안에 두었습니다.
      * MemberDTO 내부에 있는 `Request Class` 를 통해 사용자 요청 데이터를 받아오는 DTO로 활용하도록 변경했습니다.
      * Validation 관련 Annotation을 Member에서 MemberDTO로 이전했습니다.
        * 이와 관련해서는 다양한 의견들(Domain과 DTO에 둘다 Validation을 적용 해야한다 등)이 있는 것으로 보이는데, 우선 데이터를 주고 받는 객체인 DTO에 Validation을 적용시켜두고 Entity는 깔끔하게 관리가 가능한 형태로 두었습니다.
      * MemberDTO가 기존의 LoginForm 역할을 대신하기 때문에 LoginForm은 제거되었습니다.
  * 참고로 이번 PR은 리팩토링이 주가 아니기 때문에 당장 논의된 형태만 먼저 적용하였습니다. 

### 로그인 응답 처리 관련
> 공격자에게 정보가 될 수 있기 때문입니다.
  * 우선 같은 응답을 반환하도록 변경하였는데, 더 나은 구조가 있는지 고민 중입니다.

### 인증 방식 변경 및 URI 변경
  * 세션 인증 방식 -> Spring Security + JWT 토큰을 이용한 인증 방식으로 변경
  * `/member` 하위에 `/auth` 를 통해 토큰을 발급 받을 수 있도록 변경
  * 절차
    1. `/member/auth` 를 통해 아이디와 비밀번호 정보를 넘겨준다.
    2. 서버로부터 Token을 발급받는다.
    3. 인증 또는 권한이 필요한 URI로 접근할 때 Header의 `Authorization`을 통해 Bearer 타입으로 토큰을 넘겨준다.
    4. 서버는 받은 토큰을 통해 사용자를 인증하고 권한에 따라 접근을 허용한다.
  * Role
    * 아직 `Role`에 대한 내용은 논의된 바가 없기 때문에 틀만 구현 되어있습니다.
    * 회원가입을 한 모든 사용자에게 `USER` Role을 부여하고 필요에 따라 다른 Role을 추가할 수 있습니다.
    * Role에 따라 URI 별 접근을 통제할 수 있습니다.
    * Role 구성 방식에 대한 구체적인 스펙이 정해져야 구현 할 수 있을 것으로 보입니다.
      * 예시
        * USER / ADMIN처럼 Enum (String) Type으로 부여 하는 방식 (현재 적용된 방식)
        * 다중 Role 부여 방식 (Set Type)
    * 관련 견해 
      * 현재 프로젝트 규모로 봤을 때 다중 Role은 불필요할 것 같아 USER / ADMIN 정도로 구분하는 것이 괜찮아 보입니다.

## 그 외 변경 사항
### 테스트 코드
  * 로그인 관련 테스트 코드 추가
  * `org.assertj.core.api.Assertions` 으로 변경
    * `assertThat` 을 사용하는 방식이 가독성 측면에서 더 좋기 때문에 변경하였습니다.

### 예외 추가
  * Unique 한 데이터가 필요한 상황에 2개 이상이 들어올 경우에 대한 예외(`NonUniqueResultException`) 추가

## 논의가 필요한 부분
  * Role 방식 구체화
  * JWT Token을 암호화하는 SECRET Key 값 설정과 Key 관리 방법
  * Refresh Token의 필요 여부와 로그아웃 방식

B2SIC committed:
REFACTOR: Remove unused settings

namsic commented on src/main/java/harmony/dev/harmonyserver/DTO/MemberDTO.java:
```diff
@@ -0,0 +1,60 @@
+public class MemberDTO {
+
+    @Getter
+    @NoArgsConstructor
+    @AllArgsConstructor
```
컨트롤러에서는 리플렉션을 통해 값을 주입하기 때문에 기본 생성자만 있으면 될 것 같은데, 다른 생성자를 사용하는 상황이 있나요?
필요한 경우가 아니라면 어노테이션을 제거하는 것이 좋겠습니다.

DTO는 이름대로 데이터의 전달을 위한 객체이니만큼, from과 toEntity를 통한 변경 외에 DTO의 값을 직접 설정하거나 조작하는 상황은 피해야 하지 않나 하는 생각이 듭니다.

B2SIC replied:
동의합니다.
`@AllArgsConstructor`은 제거되는 것이 좋을 것으로 판단됩니다.

namsic replied:
`@NoArgsConstructor`는 남겨놓는 이유가 무엇인가요?
별도로 생성자를 명시하지 않으면 `Default Constructor(==NoArgsConstructor`가 자동 생성되지 않나요?

B2SIC replied:
여기에서는 ObjectMapper에서 기본 생성자를 반드시 필요로 하기 때문에 강조해두고자 넣어둔 코드긴 합니다.
없어도 기본 생성자가 만들어지기 때문에 문제가 되지 않습니다.

다만 무분별한 객체 생성을 방지하기 위해서
`@NoArgsConstructor(access = AccessLevel.PROTECTED)` 처럼 AccessLevel을 설정하는 것도 좋은 방법이지 않을까 생각합니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/Exception/ExceptionAdvisor.java:
```diff
@@ -45,6 +46,16 @@ protected ResponseDTO handleHttpMediaTypeNotSupportedException(HttpMediaTypeNotS
                           .build();
     }
 
+    @ExceptionHandler(NonUniqueResultException.class)
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    protected ResponseDTO handleNonUniqueResultException(NonUniqueResultException ex) {
```
500번대 에러와 관련한 의견입니다.

일반적인 시나리오에서 INTERNAL_SERVER_ERROR는 발생할 일도 없고, **발생해서는 안 되는** 에러입니다. 500번대 에러가 발생했다는 것은 그 자체로 설계나 구현을 변경해야 하는 상황임을 의미합니다.
예를 들어, Unique field인 userid나 phoneNumber는 NonUniqueResult가 절대로 나올 수 없습니다. 그럼에도 해당 예외가 발생하는 경우, DB 설정이 잘못되었거나 서버 구현이 잘못되었을 테니 관련 내용을 확인하고 수정해 해당 예외가 발생하지 않도록 해야 합니다.

따라서 INTERNAL_SERVER_ERROR는 DTO를 통해 관리할 필요가 없다고 생각합니다.
처리되지 않은 예외의 경우 자동으로 INTERNAL_SERVER_ERROR status와 함께 응답이 가기 때문이고, 이러한 예외는 발생하지 않도록 설계하는 것이 더 중요하기 때문입니다.
만약 정상 시나리오에서 NonUniqueResultException이 발생할 수 있는 경우, 409번 등 client error code를 사용하는 것이 좋겠습니다.

B2SIC replied:
말씀하신 내용에 전부 동의합니다.
`NonUniqueResultException`과 비슷한 케이스로 코드 구현을 하는 과정에서 늘 들던 의문이 있었는데, 멤버 조회 결과가 2건 이상이 나올 수 없고 나와선 안되는 케이스에 대한 예외 처리였습니다.
초기 에러 코드를 설계하는 과정에서 이 예외 처리의 필요성이 언급된 적이 있어서 추가했던 것으로 기억하는데, 서버 내부에서 발생한 오류는 발생하지 않도록 해야겠지만 이에 대한 예외 처리를 하는 것보다 INTERNAL SERVER ERROR로 던지는 것이 맞는 것 같습니다.

`NonUniqueResultException`과 함께 MemberService의 getMember() 함수 내 결과 값이 2개 이상인 경우에 대한 처리 코드를 제거하겠습니다.
또한 앞으로 이러한 케이스에 대한 예외 처리는 제외하겠습니다.

관련해서 멤버 조회 결과의 개수는 1 또는 0 밖에 존재할 수 없는데, 현재 타입을 List<Member>로 두어 빈 리스트, 1개의 결과가 담긴 리스트로 반환하고 있습니다. Member 객체를 JSON으로 변환한 값 또는 결과가 없다면 Null을 던지는 방법은 어떤지에 대한 의견도 함께 주시면 감사하겠습니다.

namsic replied:
> 관련해서 멤버 조회 결과의 개수는 1 또는 0 밖에 존재할 수 없는데, 현재 타입을 List로 두어 빈 리스트, 1개의 결과가 담긴 리스트로 반환하고 있습니다. Member 객체를 JSON으로 변환한 값 또는 결과가 없다면 Null을 던지는 방법은 어떤지에 대한 의견도 함께 주시면 감사하겠습니다.

Repository level에서는 Optional을 return하도록 하고, json 변환 시에는 null을 사용하면 될 것 같습니다.

B2SIC replied:
네 커밋에 반영해보겠습니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/service/MemberService.java:
```diff
@@ -63,4 +65,24 @@ private void validateBeforeSignupMember(Member member) {
+
+    public Map<String, String> login(String userId, String password) {
+        List<Member> userList = memberRepository.findByUserId(userId);
+        BusinessException e = new BusinessException();
+
+        if (userList.isEmpty()) {
+            e.add(ExceptionSummary.builder()
+                            .message("Login Fail")
+                            .build());
+            e.peekaboo();
+        }
+        Member findUser = userList.get(0);
+        if (!findUser.getPassword().equals(password)) {
+            e.add(ExceptionSummary.builder()
+                            .message("Login Fail")
+                            .build());
+            e.peekaboo();
+        }
```
위 경우 예외가 발생하지 않아도 매 호출마다 새 객체가 생성되기 때문에 불필요한 GC가 발생합니다.

아래는 실행해 보지 않은 코드이므로 확인이 필요합니다. 단순 참고용으로 봐 주길 바랍니다.
```java
        if (userList.isEmpty()) {
            (new BusinessException(
                ExceptionSummary.builder()
                    .code("NotExist")
                    .message("Login Fail")
                    .build()
            )).peekaboo();
            /* or
            BusinessException e = new BusinessException(
                ExceptionSummary.builder()
                    ...
            );
            e.peekaboo();
             */
        }
```

B2SIC replied:
좋은 방법인 것 같습니다.
다음 커밋에 반영하겠습니다.

namsic reviewed (commented):
# DTO
- 현재와 같은 형태가 맞는 것 같습니다. 지적해 주어 고맙습니다.
  - 각 DTO는 나중에 이름을 변경하는 것이 좋겠습니다. 아래 두 가지 정도가 생각나는데, 무엇이 더 나은지는 나중에 결정해도 될 것 같습니다.
    1. API 중심: Request 대신 SignupRequest
    => LoginRequest와 일관성 유지, 어떤 요청/응답의 DTO인지 쉽게 파악 가능
    2. 내용 중심: Request 대신 All, Response 대신 PublicInformation, LoginRequest 대신 Account
    => 여러 API에서 동일 DTO 공유하는 경우 유용할 것으로 보임, 예를 들어, userId와 password 조합은 Login뿐만 아니라 id/pw찾기, 회원정보 수정 등에서도 활용 가능한 것이 아닌지?

# Spring Security / JWT
- 배경 지식이 없어 현재 리뷰 불가합니다. 현재 PR과 프론트쪽 코드를 보면서 공부할 계획입니다. 우선은 직접 판단해서 처리해 주기를 바랍니다.
- Role 관련: 현재 프로젝트에서는 아마 단순한 Role으로 충분할 것 같은데, 실무에서 관련 지식이나 경험이 있으면 도움이 많이 되는 것으로 알고 있습니다. 꼭 필요한 role이 아니어도 연습삼아 구현해 보는 것도 나쁘지 않다고 생각합니다. 원하는 대로 진행 바랍니다.

# 기타
기타 설계와 관련해서 논의가 필요한 부분은 서버 외 다른 팀원도 함께 의논해야 할 것 같습니다.

B2SIC committed:
REFACTOR: Reflect review
Remove AllArgsConstructor Annotation in MemberDTO
Remove NonUniqueResultExceiption and related codes
Change object calling type

B2SIC commented:
## 의견
> API 중심: Request 대신 SignupRequest
=> LoginRequest와 일관성 유지, 어떤 요청/응답의 DTO인지 쉽게 파악 가능
내용 중심: Request 대신 All, Response 대신 PublicInformation, LoginRequest 대신 Account
=> 여러 API에서 동일 DTO 공유하는 경우 유용할 것으로 보임, 예를 들어, userId와 password 조합은 Login뿐만 아니라 id/pw찾기, 회원정보 수정 등에서도 활용 가능한 것이 아닌지?

이름 변경이 필요하다는 말씀에 동의합니다.
제시해준 두 방식 모두 괜찮은 것 같은데, 클래스를 최대한 재사용이 가능한 형태로 구성하면 좋지 않을까 생각합니다.

## 리뷰 내용 반영
* 불필요한 생성자 Annotation 제거
* NonUniqueResultException 제거 및 관련 예외 처리 코드 제거
* 예외 객체 생성 방식 변경

B2SIC committed:
REFACTOR: Reflect review
Add AccessLevel at NoArgsConstructor Annotation
Change return type for find query method(List -> Optional)
Change return type for test code

B2SIC commented:
## 리뷰 내용 반영
> Repository level에서는 Optional을 return하도록 하고, json 변환 시에는 null을 사용하면 될 것 같습니다.
* findByOptionalParameters의 리턴 타입을 `List<Member>`에서 `Optional<Member>`로 변경
  * 관련 코드 (테스트 코드 포함) 변경
* `@NoArgsConstructor`에 AccessLevel 추가 -> 임의 객체 생성 예방

## 관련 Issue
* findByOptionalParameters의 리턴 타입을 Optional로 바꾸는 과정에서 `getResultStream()`을 사용했는데, 이 메서드가 **커서 기능**을 동작 시키기 때문에 커서 유지를 위해 `@Transactional` 을 메서드에 추가하지 않으면 Exception이 발생했습니다.
  * Exception Message: `The object is already closed` + `could not advance using next()`
  * 원인은 스트림을 받아올 때 JDBC 드라이버에서 next()를 호출하여 다음 데이터를 차례로 가져오는 동작을 반복하는데, 커서 데이터가 어떤 원인(아마도 커밋)에 의해 끊어졌기 때문인 것으로 보입니다.
* 추가로 **커서 기능**이 데이터베이스 리소스를 많이 잡아먹는다는 이슈가 존재합니다.
  * 다만 findByOptionalParameters는 대용량의 데이터를 가져오는 동작이 아니긴 함.

namsic commented on src/main/java/harmony/dev/harmonyserver/repository/DefaultMemberRepository.java:
```diff
@@ -22,22 +23,25 @@ public Member save(Member member) {
-    public List<Member> findByOptionalParameters(String userId, String phoneNumber) {
+    @Transactional(readOnly = true)
+    public Optional<Member> findByOptionalParameters(String userId, String phoneNumber) {
         return em.createQuery("SELECT m FROM Member m WHERE 1 = 1"
                 + " AND (:userId is null OR m.userId = :userId)"
                 + " AND (:phoneNumber is null OR m.phoneNumber = :phoneNumber)", Member.class)
                 .setParameter("userId", userId)
                 .setParameter("phoneNumber", phoneNumber)
-                .getResultList();
+                .getResultStream().findAny();
     }
 
     @Override
-    public List<Member> findByUserId(String userId) {
+    @Transactional(readOnly = true)
+    public Optional<Member> findByUserId(String userId) {
         return findByOptionalParameters(userId, null);
     }
 
     @Override
-    public List<Member> findByPhoneNumber(String phoneNumber) {
+    @Transactional(readOnly = true)
+    public Optional<Member> findByPhoneNumber(String phoneNumber) {
         return findByOptionalParameters(null, phoneNumber);
     }
```
현재 기준으로 Optional 입력을 받는 설계는 가능한 하지 않을 계획이므로
`findByOptionalParameters()` 자체를 제거하는 것은 어떤가요?

이 경우 이전에 사용했던 아래와 같은 형태의 구현으로 돌아가야 하지 않나 싶습니다.
https://github.com/B2SIC/harmony-server/blob/91dc74019942d41c5047a8b16beea7ce22842b64/src/main/java/harmony/dev/harmonyserver/repository/DefaultMemberRepository.java#L23-L37

추가로 이러한 로직은 병렬 처리가 필요하지는 않은 것 같은데
이 경우에도 아래처럼 List를 바로 다루는 것보다
stream으로 변환한 뒤 findAny()를 하는 것이 더 효율적인가요?
```java
// 정확한 exception 이름이나 Optional 사용법은 잊어버렸음
if (result.isEmpty()) return new Optional.Null;
if (result.size() == 1) return new Optional(result.get(0));
throw new NonUniqueKeyException;
```

B2SIC replied:
> 현재 기준으로 Optional 입력을 받는 설계는 가능한 하지 않을 계획이므로
findByOptionalParameters() 자체를 제거하는 것은 어떤가요?

관련 Issue가 별도로 존재해서 API URI 분리와 함께 처리하면 좋을 것 같습니다.
이전에 구현해둔 코드가 존재하기도 해서 그대로 사용해도 문제 없어보입니다.

> 추가로 이러한 로직은 병렬 처리가 필요하지는 않은 것 같은데
이 경우에도 아래처럼 List를 바로 다루는 것보다
stream으로 변환한 뒤 findAny()를 하는 것이 더 효율적인가요?

stream을 사용한 이유는 변환 대상이 많아야 1개이기 때문에 효율성보다는 코드 가독성을 높이는 방식이 더 가치 있다고 생각했기 때문입니다.
예) stream().findAny() 를 통해 Optional 객체로 반환할 경우 data 값 전달을 한 줄로 처리가 가능했습니다.
`memberService.getMembers(params).map(MemberDTO.Response::from).orElse(null)`

그런데 과거 find가 분리된 코드를 보니 `getResultList()` 로 결과를 리스트 타입으로 받아와서 `stream().findAny()`를 통해 처리해도 괜찮을 것 같습니다. 이렇게 할 경우 `@Transactional`은 없어도 될 것으로 생각됩니다.
이 내용도 findByOptionalParameters()를 제거할 때 같이 변경 되는 사항이니 다음 Issue로 넘겨서 처리해도 될 것 같습니다.

namsic replied:
> 효율성만 놓고 봤을 때는 stream 변환 대상이 많아질수록 비효율적인 코드가 될 것으로 보입니다.

- Stream은 여러 데이터에 대한 병렬 처리가 가능하기 때문에 오히려 많은 양의 데이터를 처리해야 하는 상황일수록 List => Stream 변환 cost를 감수하고 Stream을 사용하는 것이 더 효율적인 것으로 알고 있는데, 아닌가요?

> 그런데 과거 find가 분리된 코드를 보니 getResultList() 로 결과를 리스트 타입으로 받아와서 stream().findAny()를 통해 처리해도 괜찮을 것 같습니다. 이렇게 할 경우 `@Transactional`은 없어도 될 것으로 생각됩니다.

- findAny()를 활용하면 Entity 설정이 항상 완벽한 상태이리라 기대하고 NonUniqueKey에 대한 핸들링은 별도로 하지 않는 형태인가요?
```java
default Stream<X> getResultStream() {
    return getResultList().stream();
}
```
- `getResultStream()` 내부 구현에서 `getResultList()` 호출 뒤 변환하는 방식을 사용하고 있으므로 효율성 및 동작 방식은 동일할 것으로 보입니다.

B2SIC replied:
> Stream은 여러 데이터에 대한 병렬 처리가 가능하기 때문에 오히려 많은 양의 데이터를 처리해야 하는 상황일수록 List => Stream 변환 cost를 감수하고 Stream을 사용하는 것이 더 효율적인 것으로 알고 있는데, 아닌가요?
* 적어주신 내용은 맞습니다, 비효율을 적었던 의도는 getResultStream을 사용했을 때 커서 기능으로 인한 오버헤드를 우려했던건데, 다시 읽어보니 질문의 의도가 그것이 아니었던 것 같아 내용을 수정 했습니다. 지적 감사합니다.

> findAny()를 활용하면 Entity 설정이 항상 완벽한 상태이리라 기대하고 NonUniqueKey에 대한 핸들링은 별도로 하지 않는 형태인가요?
* 맞습니다, 앞선 리뷰에서 namsic님의 의견에 따라 예외 처리를 제거했던 부분입니다.
  > 예를 들어, Unique field인 userid나 phoneNumber는 NonUniqueResult가 절대로 나올 수 없습니다. 그럼에도 해당 예외가 발생하는 경우, DB 설정이 잘못되었거나 서버 구현이 잘못되었을 테니 관련 내용을 확인하고 수정해 해당 예외가 발생하지 않도록 해야 합니다.

> getResultStream() 내부 구현에서 getResultList() 호출 뒤 변환하는 방식을 사용하고 있으므로 효율성 및 동작 방식은 동일할 것으로 보입니다.
* 저도 의아한 부분이라 직접 테스트를 해봤는데 JPA에서 발생하는 [버그](https://stackoverflow.com/questions/40775697/streaming-query-results-closes-prematurely-spring-data-jpa-and-hibernate)로 보입니다.
  * 토큰 인증을 하는 부분에서 `.getResultList().stream().findAny()` 으로 결과를 받아오면 `@Transactional` 없이 정상 동작하고 `.getResultStream().findAny()` 으로 결과를 받아오면 `@Transactional` 이 없을 경우 위에서 언급했던 Exception이 발생합니다.

namsic replied:
> 맞습니다, 앞선 리뷰에서 namsic님의 의견에 따라 예외 처리를 제거했던 부분입니다.

이와는 상황이 약간 다릅니다. 참고 바랍니다.

- 이전 구현에서 DB 설정이 잘못된 경우
  1. findBy 결과가 둘 이상이면 NoneUnique 예외 throw
  2. ExceptionAdviser에서 별도로 응답을 DTO로 wrapping하지는 않음
  => 일반적인 상황에서 절대 발생하지 않는 예외이며, 발생 확인되는대로 수정해야 하기 때문

- 현재 구현에서 DB 설정이 잘못된 경우
  1. findBy 결과가 둘 이상이면 임의의 값 하나 선택해서 반환
  => 일반적 상황뿐만 아니라 문제가 있는 상황에서도 예외 발생하지 않음
  2. 겉으로 드러나는 문제 없이 로직 진행
  => 어떤 side-effect 발생할지 예측 불가
  => 근본 원인은 DB 설정인데 다른 곳에서 에러 발생 시 원인 파악에 어려움 겪을 수 있음

B2SIC replied:
그렇다면 문제 상황이 발생할 경우 로깅을 통해 확인 가능한 형태로 남기는 방법이 좋을까요?
굳이 에러 코드를 통해 클라이언트 측에 전달 할 필요는 없어보입니다.

namsic reviewed (approved):
LGTM

B2SIC committed:
Add Slf4j Logger
Change return Type for findByOptionalParameters
Remove Transactional annotation
